src/toolbox/shamir.rs

```diff
@@ -116,48 +94,70 @@ fn evaluate_lagrange(x: Scalar, xis: &Vec<Scalar>, yis: &Vec<Scalar>) -> Result<
 
     /// Given a secret and some other information, calculate a set of shares using Shamir's Secret Sharing.  We don't
     /// expect this code to ever be called by our application, but it's useful for testing.
-    pub fn share<R: RngCore + CryptoRng>(secret: &Scalar, nr_of_shares: usize, threshold: usize, rng: &mut R) -> SecretShare {
+    fn share(&mut self, secret: &Scalar, nr_of_shares: usize) -> Vec<Option<Scalar>> {
 
         // first, select random coefficients for each term (less the first) in the polynomial
         let mut coefficients: Vec<Scalar> = Vec::new();
         let my_secret = secret.clone();
         coefficients.push(my_secret);  // the coefficient for x^0 is the secret
-        for _ in 1..(threshold) {
-            coefficients.push(Scalar::random(rng));
+        for _ in 1..(self.threshold) {
+            coefficients.push(Scalar::random(&mut self.rng));
         }
 
         // now, calculate points on the line as our shares.
         // we encode the x value as the index of the vec
-        let mut shares: Vec<Scalar> = Vec::new();
+        let mut shares: Vec<Option<Scalar>> = Vec::new();
         for i in 0..nr_of_shares {
             let x = Scalar::from((i+1) as u64);
             let y = calc_polynomial(&coefficients, &x);
-            shares.push(y);
+            shares.push(Some(y));
         }
 
-        SecretShare {secret: my_secret, shares, threshold}
+        shares
     }
 
     /// Given some shares and the secret, compute a valid set of additional shares.  Note that the new shares have
     /// NO GUARANTEE of being derived from the same polynomial as the provided ones.  We pick an all-new polynomial
     /// which fits the provided points.  Camenish refers to this function as cmpl_Gamma in his thesis.
     ///
-    /// The sparse_shares vector should follow the "index + 1 = x" convention, but is allowed to contain Option values
+    /// The sparse_shares vector should follow the "index + 1 = x" convention, but is allowed to contain None values
     /// to represent missing shares.  DO NOT include the secret in sparse_shares, as it is passed in separately.
-    pub fn complete<R: CryptoRng + RngCore>(secret: Scalar, threshold: usize, sparse_shares: &Vec<Option<Scalar>>, rng: &mut R) -> Result<SecretShare, String> {
+    fn complete(&mut self, secret: &Scalar, sparse_shares: &Vec<Option<Scalar>>) -> Result<Vec<Option<Scalar>>, String> {
+        if self.threshold == 0 {
+            // Threshold of 0 should be impossible, since it's going to at least have a single point (the secret)
+            return Err(String::from("Threshold of zero is invalid for Shamir's Secret Sharing"));
+        } else if self.threshold == 1 {
+            // If threshold is 1, we're reconstructing a single point.  The polynomial is y = secret.
+            return Ok(sparse_shares.iter().map(|s| match s { Some(x) => Some(*x), None => Some(*secret)}).collect())
```

Comment:
This seems not always the case. If threshold is 1, and I give a vector with multiple points that are not equal to the secret, it should give me an error. Basically, you have to validate that with complete the threshold is indeed equal to the number of `Some` in the `sparse_shares` vector (+/- 1). If more scalars are given, you have to check if they are indeed on a valid polynomial, otherwise you also have to give an error.

Reply:
The special case was intended to speed things up, but I can remove it.